dvc/repo/experiments/__init__.py

```diff
@@ -563,22 +563,26 @@ def _checkpoint_callback(rev, executor, unchanged, stages):
                     result[rev] = {exp_rev: exp_hash}
 
             checkpoint_func = partial(_checkpoint_callback, rev, executor)
-            executor.reproduce(
+
+            exp_hash = executor.reproduce(
                 executor.dvc_dir,
                 cwd=executor.dvc.root_dir,
                 checkpoint=True,
                 checkpoint_func=checkpoint_func,
                 **executor.repro_kwargs,
             )
 
-            # TODO: determine whether or not we should create a final
-            # checkpoint commit after repro is killed, or if we should only do
-            # it on explicit make_checkpoint() signals.
-
-        # NOTE: our cached GitPython Repo instance cannot be re-used if the
-        # checkpoint run was interrupted via SIGINT, so we need this hack
-        # to create a new git repo instance after checkpoint runs.
-        del self.scm
+            # NOTE: GitPython Repo instances cannot be re-used after
+            # process has received SIGINT or SIGTERM, so we need this hack
+            # to re-instantiate git instances after checkpoint runs. See:
+            # https://github.com/gitpython-developers/GitPython/issues/427
+            del self.repo.scm
+            del self.scm
```

Comment:
The leftover `BrokenPipeError` from the initial checkpoints demo was because we were previously only re-instantiating `experiments.scm` after checkpoint runs (and not parent `repo.scm`).

As previously discussed, the need for this hack may be avoided by moving to dulwich or pygit, but checkpoints is currently the only place in DVC where we should be doing any git operations after SIGINT/SIGTERM, so this workaround works for now.